### PR TITLE
Proposed simplification.

### PR DESCRIPTION
**Pull request by bwplotka**
I did not fix all the things, just the flow I mentioned:

* Node is just data structure. NO magic methods. It allows flexibility and simplicity.
* I simplified a node a lot
* add method replaced and simplified as well (why you had bool return arg? Why you have metas map?)

Let me know if that makes sense. Feel free to merge, test and then I can then review  all.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/block/fetcher.go [hunk 1]
```diff
@@ -419,56 +419,51 @@ type DeduplicateFilter struct {
 
 // NewDeduplicateFilter creates DeduplicateFilter.
 func NewDeduplicateFilter() *DeduplicateFilter {
-	return &DeduplicateFilter{
-		DuplicateIDs: []ulid.ULID{},
```
Comment (bwplotka) [inline pkg/block/fetcher.go:423]: No need for this

File: pkg/block/node.go [hunk 2]
```diff
@@ -8,50 +8,26 @@ import (
 	"github.com/thanos-io/thanos/pkg/block/metadata"
 )
 
-type addNode func(root *Node, metas map[ulid.ULID]*metadata.Meta, node *Node) bool
-
-// Node type implements a node of a tree.
+// Node type represents a node of a tree.
 type Node struct {
-	ID       ulid.ULID
-	Value    *metadata.Meta
```
Comment (bwplotka) [inline pkg/block/node.go:16]: This was unnecessary complex and there was duplicate of information

File: pkg/block/fetcher.go [hunk 3]
```diff
@@ -419,56 +419,51 @@ type DeduplicateFilter struct {
...
 	for _, id := range duplicateULIDs {
 		if metas[id] != nil {
 			f.DuplicateIDs = append(f.DuplicateIDs, id)
 		}
+		// TODO(khyati): Increment synced metrics with proper state.
 		delete(metas, id)
 	}
 }
 
-func addNodeToRoot(startNode *Node, metas map[ulid.ULID]*metadata.Meta, newNode *Node) bool {
```
Comment (bwplotka) [inline pkg/block/fetcher.go:462]: no idea why you used meta

File: pkg/block/fetcher.go (hunk 1 above)
Comment (khyatisoneji) [reply pkg/block/fetcher.go:423]: yes right now its not required, but it would be required when we want to mark the filtered nodes as deleted

Comment (khyatisoneji) [reply pkg/block/fetcher.go:423]: I was initializing the DuplicateIds slice in the constructor, is that not required?

Comment (bwplotka) [reply pkg/block/fetcher.go:423]: No (:

Comment (khyatisoneji) [reply pkg/block/fetcher.go:423]: It gives nullpointerexception if we access Children of node. So adding constructor for node

File: pkg/block/fetcher.go [hunk 4]
```diff
@@ -419,56 +419,51 @@ type DeduplicateFilter struct {
...
 // Filter filters out duplicate blocks that can be formed
 // from two or more overlapping blocks that fully submatches the source blocks of the older blocks.
 func (f *DeduplicateFilter) Filter(metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) {
-	root := NewNode(ulid.MustNew(uint64(0), nil), nil)
+	root := &Node{}
 
-	metaSlice := []*metadata.Meta{}
 	for _, meta := range metas {
-		metaSlice = append(metaSlice, meta)
+		root.Children = append(root.Children, &Node{Meta: *meta})
```
Comment (khyatisoneji) [inline pkg/block/fetcher.go:431]: Adding all nodes as children on root and then adding them again goes into stackoverflow exception, so sorting it by adding all nodes to slice and sorting slice.

**Pull request merged**